yb-voyager/src/tgtdb/yugabytedb.go

```diff
@@ -491,21 +531,49 @@ func (yb *TargetYugabyteDB) importBatch(conn *pgx.Conn, batch Batch, args *Impor
 		}
 	}()
 
-	// Check if the split is already imported.
-	var alreadyImported bool
-	alreadyImported, rowsAffected, err = yb.isBatchAlreadyImported(tx, batch)
+	// using the conn on which the transaction is executing which should ensure -
+	// all DB operations inside copyBatchCore will be a part of the transaction
+	rowsAffected, err = yb.copyBatchCore(tx.Conn(), batch, args)
+	return rowsAffected, err
+}
+
+func (yb *TargetYugabyteDB) importBatchFast(conn *pgx.Conn, batch Batch, args *ImportBatchArgs) (int64, error) {
+	return yb.copyBatchCore(conn, batch, args)
+}
+
+func (yb *TargetYugabyteDB) copyBatchCore(conn *pgx.Conn, batch Batch, args *ImportBatchArgs) (int64, error) {
+	// 1. Open the batch file
+	file, err := batch.Open()
+	if err != nil {
+		return 0, fmt.Errorf("open file %s: %w", batch.GetFilePath(), err)
+	}
+	defer file.Close()
+
+	// 2. setting the schema so that COPY command can acesss the table
+	// Q: If we set the schema for this batch on this conn, will it impact others using the same conn from pool later?
+	yb.setTargetSchema(conn)
+
+	// 3. Check if the split is already imported.
+	alreadyImported, rowsAffected, err := yb.isBatchAlreadyImported(conn, batch)
 	if err != nil {
 		return 0, err
 	}
 	if alreadyImported {
 		return rowsAffected, nil
 	}
 
-	// Import the split using COPY command.
+	// 4. Import the batch using COPY command.
 	var res pgconn.CommandTag
-	copyCommand := args.GetYBCopyStatement()
+	var copyCommand string
+
+	// TODO: maybe move this check into the args struct methods and just call GetYBCopyStatement()
```

Comment:
Your intuition is right. You probably shouldn't be doing this check again, since you've already done an if-else check before. Also, fastPath or not is not relevant at this layer (i.e. this function where we're interested in just executing copy command)
How about: 
`copyBatchCore` has an arg `inTxn bool` . In case of fast path, you pass false, in case of normal path, you pass true. 
`copyBatchCore` chooses which type of statement it wants based on value of `inTxn`.

Reply:
Thinking of this

```
func (args *ImportBatchArgs) GetYBCopyStatement() string {
	options := args.copyOptions()
	if !args.CanUseFastPath() { 
		// fast path on DB side not enabled if ROWS_PER_TRANSACTION is set in COPY
		options = append(options, fmt.Sprintf("ROWS_PER_TRANSACTION %v", args.RowsPerTransaction))
	}

	columns := ""
	if len(args.Columns) > 0 {
		columns = fmt.Sprintf("(%s)", strings.Join(args.Columns, ", "))
	}

	return fmt.Sprintf(`COPY %s %s FROM STDIN WITH (%s)`, args.TableNameTup.ForUserQuery(), columns, strings.Join(options, ", "))
}
```

Reply:
done